domain/modelconfig/service/service.go

```diff
@@ -69,29 +82,71 @@ type WatcherFactory interface {
 type Service struct {
 	defaultsProvider ModelDefaultsProvider
 	modelValidator   config.Validator
+	ctrlSt           ControllerState
 	st               State
 }
 
 // NewService creates a new ModelConfig service.
 func NewService(
 	defaultsProvider ModelDefaultsProvider,
 	modelValidator config.Validator,
+	ctrlSt ControllerState,
 	st State,
 ) *Service {
 	return &Service{
 		defaultsProvider: defaultsProvider,
 		modelValidator:   modelValidator,
+		ctrlSt:           ctrlSt,
 		st:               st,
 	}
 }
 
+func (s *Service) configHandlers(ctx context.Context, modelUUID model.UUID) ([]handlers.ConfigHandler, error) {
+	defaults, err := s.defaultsProvider.ModelDefaults(ctx)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot load model defaults")
+	}
+	return []handlers.ConfigHandler{
+		handlers.SecretBackendHandler{
+			Defaults:     defaults,
+			BackendState: s.ctrlSt,
+			ModelUUID:    modelUUID,
+		},
+	}, nil
+}
+
+func (s *Service) runOnLoadHandlers(ctx context.Context, modelUUID model.UUID, stConfig map[string]string) error {
+	cfgHandlers, err := s.configHandlers(ctx, modelUUID)
+	if err != nil {
+		return fmt.Errorf("cannot set up config handlers: %w", err)
+	}
+	for _, h := range cfgHandlers {
+		additionalCfg, err := h.OnLoad(ctx)
+		if err != nil {
+			return fmt.Errorf("error running onLoad config handler %s: %w", h.Name(), err)
+		}
+		for k, v := range additionalCfg {
+			stConfig[k] = v
+		}
+	}
+	return nil
+}
+
 // ModelConfig returns the current config for the model.
 func (s *Service) ModelConfig(ctx context.Context) (*config.Config, error) {
 	stConfig, err := s.st.ModelConfig(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("getting model config from state: %w", err)
 	}
 
+	modelUUID, _, err := s.st.GetModelInfo(ctx)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot load model info")
+	}
+	if err := s.runOnLoadHandlers(ctx, modelUUID, stConfig); err != nil {
+		return nil, errors.Annotate(err, "cannot process model config")
+	}
+
 	agentVersion, err := s.st.AgentVersion(ctx)
```

Comment:
Can you please re-implement AgentVersion as a modelconfig handler, just to prove it out more and to ensure we are being consistent.

Reply:
The PR was quite large already and I was limiting the scope to just one change. There's at least 2 more implementations of this pattern still to do:
- agent version
- model default space

The intent was to do these in a follow up. If you're happy to review, I can do the agent version one here.

Reply:
these can be done in a separate pr